vtex/faststore
# Chore: Adds styles layers to help with CSS specificity issues

## What's the purpose of this pull request?

- Adds styles layers to deal with CSS specificity issues

## How it works?

We face a CSS specificity issue when using the overriding component and creating a custom component using the FS UI component. We are setting an order for customization CSS specificity through css cascade layers.

|Before|
|-|
|<img width="1262" alt="image" src="https://github.com/vtex/faststore/assets/3356699/7be68018-18c8-47b9-8a8c-33ea5b62dc0c">|

Now the priority should be:
`Users custom styles**` > `theme` > `components (default ui styles)`

** un-layered styles have the highest priority

|w/ css layers|
|-|
|<img width="1163" alt="image" src="https://github.com/vtex/faststore/assets/3356699/434c6b6b-2223-4dd2-ba21-9abe01c4d1d7">|

The cascade layers feature seems to have good coverage (91.23%) support by the browsers. So we've decided it would be a practical approach for this issue.

## How to test it?

You can go through this [guide](https://faststore-site-git-chore-adds-styles-layers-fs-1185-faststore.vercel.app/docs/customizing/components#customizing-a-new-component-when-using-override), and play with the custom stylesheet. Check if the custom stylesheet has a higher priority when setting your styles.

1. Try it with the current starter repo version (You won't see changes in the button's borders nor padding)
2. Then try on this [starter branch](https://github.com/vtex-sites/starter.store/pull/114).  (You will be able to see all the changes applied)

## References

- [Cascade Layers](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Cascade_layers)
- https://caniuse.com/css-cascade-layers
